docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+	return &StoreKVPairWriteListener{
+		writer: w,
+		marshaler: m,
+	}
+}
+
+// OnWrite satisfies the WriteListener interface by writing length-prefixed protobuf encoded StoreKVPairs
+func (wl *StoreKVPairWriteListener) OnWrite(storeKey types.StoreKey, key []byte, value []byte) {
+	kvPair := new(types.StoreKVPair)
+	kvPair.StoreKey = storeKey.Name()
+	kvPair.Key = key
+	kvPair.Value = value
+	if by, err := wl.marshaler.MarshalBinaryLengthPrefixed(kvPair); err == nil {
+		wl.writer.Write(by)
+	}
+}
+```
+
+### ListenKVStore
+
+We will create a new `Store` type `listenkv.Store` that the `MultiStore` wraps around a `KVStore` to enable state listening.
+We can configure the `Store` with a set of `WriteListener`s which stream the output to specific destinations.
+
+```go
+// Store implements the KVStore interface with listening enabled.
+// Operations are traced on each core KVStore call and written to any of the
+// underlying listeners with the proper key and operation permissions
+type Store struct {
+	parent    types.KVStore
+	listeners []types.WriteListener
+	parentStoreKey types.StoreKey
+}
+
+// NewStore returns a reference to a new traceKVStore given a parent
+// KVStore implementation and a buffered writer.
+func NewStore(parent types.KVStore, psk types.StoreKey, listeners []types.WriteListener) *Store {
+	return &Store{parent: parent, listeners: listeners, parentStoreKey: psk}
+}
+
+// Set implements the KVStore interface. It traces a write operation and
+// delegates the Set call to the parent KVStore.
+func (s *Store) Set(key []byte, value []byte) {
+	types.AssertValidKey(key)
+	s.parent.Set(key, value)
+	s.onWrite(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (s *Store) Delete(key []byte) {
+	s.parent.Delete(key)
+	s.onWrite(key, nil)
+}
+
+// onWrite writes a KVStore operation to all of the WriteListeners
+func (s *Store) onWrite(key, value []byte) {
+	for _, l := range s.listeners {
+		l.OnWrite(s.parentStoreKey, key, value)
+	}
+}
+```
+
+### MultiStore interface updates
+
+We will update the `MultiStore` interface to allow us to wrap a set of listeners around a specific `KVStore`.
+Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer.
+
+```go
+type MultiStore interface {
+	...
+
+	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
+	ListeningEnabled(key StoreKey) bool
+
+	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
+	// It appends the listeners to a current set, if one already exists
+	SetListeners(key StoreKey, listeners []WriteListener)
+}
+```
+
+```go
+type CacheWrap interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(storeKey types.StoreKey, listeners []WriteListener) CacheWrap
```

Comment:
listeners here listen to changes in the cache, even if they are never written to the underlying store?

Reply:
My thinking was that this would work similarly to how the tracekv is used in the cache layer. The listenkv store will be wrapped around the parent kvstore in the cache kvstore (a la https://github.com/cosmos/cosmos-sdk/blob/master/store/iavl/store.go#L159) and then the listening would only occur when the cache layer updates are [written out to the underlying parent kvstore](https://github.com/cosmos/cosmos-sdk/blob/master/store/cachekv/store.go#L120).